lib/proxy/peer/quic/quic.go

```diff
@@ -0,0 +1,189 @@
+of the early data, as an additional layer against replay attacks; this will
+cause no further delays if the client intends to wait for the server to reply to
+the dial request. A client that wants to make use of multiplexing should take
+care to not accidentally send more than one dial request as 0-RTT in a single
+connection, to keep the effort needed to handle potential replays at a minimum.
+
+The protocol doesn't currently take advantage of early server-side data for
+non-resumed connections, so considerations around the security of "0.5-RTT" data
+are not relevant; data sent by the server as a response to the client is either
+using 0-RTT keys or is sent after the handshake is completed.
+*/
+package quic
+
+import (
+	"encoding/binary"
+	"errors"
+	"math"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/quic-go/quic-go"
+	"github.com/quic-go/quic-go/quicvarint"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	// quicMaxIdleTimeout is the arbitrary timeout after which a QUIC connection
+	// that hasn't received data is presumed to be lost to the aether.
+	quicMaxIdleTimeout = 30 * time.Second
+	// quicKeepAlivePeriod is the interval of QUIC keepalive packets sent if the
+	// connection is otherwise idle.
+	quicKeepAlivePeriod = 5 * time.Second
+
+	quicMaxReceiveWindow   = quicvarint.Max
+	quicMaxIncomingStreams = 1 << 60 // maximum allowed value as per the quic-go docs
+
+	// quicNextProto is the ALPN indicator for the current version of the QUIC
+	// proxy peering protocol.
+	quicNextProto = "teleport-peer-v1alpha"
+
+	// quicMaxMessageSize is the maximum accepted size (in protobuf binary
+	// format) for the request and response messages exchanged as part of the
+	// dialing.
+	quicMaxMessageSize = 128 * 1024
+
+	// quicTimestampGraceWindow is the maximum time difference between local
+	// time and reported time in a 0-RTT request. Clients should not keep trying
+	// to use a request after this much time has passed.
+	quicTimestampGraceWindow = time.Minute
+	// quicNoncePersistence is the shortest time for which a nonce will be kept
+	// in memory to prevent 0-RTT replay attacks. Should be significantly longer
+	// than [quicTimestampGraceWindow]. In the current implementation, nonces
+	// can be kept for at most twice this value.
+	quicNoncePersistence = 5 * time.Minute
+
+	quicDialTimeout          = 30 * time.Second
+	quicRequestTimeout       = 10 * time.Second
+	quicErrorResponseTimeout = 10 * time.Second
+)
+
+// marshalSized returns the wire encoding of the given [proto.Message] prefixed
+// by its length encoded as a little endian 32-bit integer.
+func marshalSized(m proto.Message) ([]byte, error) {
+	buf := make([]byte, 4, 4+proto.MarshalOptions{}.Size(m))
+	buf, err := proto.MarshalOptions{UseCachedSize: true}.MarshalAppend(buf, m)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if len(buf)-4 > math.MaxUint32 {
+		return nil, trace.LimitExceeded("oversized message")
```

Comment:
Is trace.LimitExceeded correct? Isn't that more related to rate-limiting than an message-size failure?

Reply:
`LimitExceeded` maps to `codes.ResourceExhausted`, which is generated when `grpc-go` deals with an oversized message. Rate limiting would probably be `codes.Unavailable` (which weirdly enough maps to `ConnectionProblemError` in `trace/trail`).